OneTimePasswordTests/KeychainTests.swift

```diff
@@ -0,0 +1,215 @@
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+//  SOFTWARE.
+//
+
+import XCTest
+import OneTimePassword
+
+let testToken = Token(
+    name: "Name",
+    issuer: "Issuer",
+    generator: Generator(
+        factor: .Timer(period: 45),
+        secret: NSData(base32String: "AAAQEAYEAUDAOCAJBIFQYDIOB4"),
+        algorithm: .SHA256,
+        digits: 8
+    )!
+)
+
+class KeychainTests: XCTestCase {
+    let keychain = Keychain.sharedInstance
+
+    func testPersistentTokenWithIdentifier() {
+        // Create a token
+        let token = testToken
+
+        do {
+            // Save the token
+            let savedToken = try keychain.addToken(token)
+
+            // Restore the token
+            do {
+                let fetchedToken = try keychain.persistentTokenWithIdentifier(savedToken.identifier)
+                XCTAssertEqual(fetchedToken, savedToken, "Token should have been saved to keychain")
+            } catch {
+                XCTFail("persistentTokenWithIdentifier(_:) failed with error: \(error)")
+            }
+
+            // Modify the token
+            let modifiedToken = Token(
+                name: "New Name",
+                issuer: "New Issuer",
+                generator: token.generator.successor()
+            )
+            do {
+                let updatedToken = try keychain.updatePersistentToken(savedToken,
+                    withToken: modifiedToken)
+                XCTAssertEqual(updatedToken.identifier, savedToken.identifier)
+                XCTAssertEqual(updatedToken.token, modifiedToken)
+            } catch {
+                XCTFail("updatePersistentToken(_:withToken:) failed with error: \(error)")
+            }
+
+            // Fetch the token again
+            do {
+                let fetchedToken = try keychain.persistentTokenWithIdentifier(savedToken.identifier)
+                XCTAssertEqual(fetchedToken?.token, modifiedToken)
+                XCTAssertEqual(fetchedToken?.identifier, savedToken.identifier)
+            } catch {
+                XCTFail("persistentTokenWithIdentifier(_:) failed with error: \(error)")
+            }
+
+            // Remove the token
+            do {
+                try keychain.deletePersistentToken(savedToken)
+            } catch {
+                XCTFail("deletePersistentToken(_:) failed with error: \(error)")
+            }
+
+            // Attempt to restore the deleted token
+            do {
+                let fetchedToken = try keychain.persistentTokenWithIdentifier(savedToken.identifier)
+                XCTAssertNil(fetchedToken, "Token should have been removed from keychain")
+            } catch {
+                XCTFail("persistentTokenWithIdentifier(_:) failed with error: \(error)")
+            }
+        } catch {
+            XCTFail("addToken(_:) failed with error: \(error)")
+        }
+    }
+
+    func testDuplicateTokens() {
```

Comment:
Length Violation: Function body should be span 40 lines or less: currently spans 70 lines